### DISCUSSION TITLE
Tried to create a custom decorator for my commands but couldn't be done.

### DISCUSSION THREAD

**Discussion by klaiderman** [python-telegram-bot/python-telegram-bot#3946] [Q&A] [answered]
Hey, I have tried to create a custom decorator for my commands, but it couldn't be done and I have no idea why!
I would be glad if someone could help(:
```python
def command(command_name):
    def decorator(func):
        async def wrapper(update: Update, context: CustomContext):
            await func(update, context)
        wrapper.is_command = True
        return wrapper 
    return decorator
```
```python
class MyBot: 
    def __init__(self):
        self.app = ApplicationBuilder().token(...).build()
        self.app.add_handler(CommandHandler('demo', self.demo_command))
        
    def run(self):
        ...
            
    @command(name='demo')
    async def demo_command(self, update, context):
        print('demo_command has been called')
        
if __name__ == '__main__':
    MyBot().run()
```
```cmd
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\telegram\ext\_application.py", line 1195, in process_update
    await coroutine
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\telegram\ext\_basehandler.py", line 153, in handle_update
    return await self.callback(update, context)
TypeError: wrapper() takes 2 positional arguments but 3 were given
```

**Comment by Poolitzer** [comment]
You give an additional argument to your wrapper function, name. You need to add this as a parameter to the function, after update and context.

**Reply by Bibo-Joshi** [reply]
Isn't the issue rather that `demo_command` has an argument `self`?

**Reply by klaiderman** [reply]
I see. But, if I want to pass arguments to the decorator itself ( `@command(name=...)`, I need to define it in the `def command(name)`. Isn't it? ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀^

**Reply by klaiderman** [reply]
> Isn't the issue rather that `demo_command` has an argument `self`?

I thought about it before this as well, but, from what I know, `self` has nothing to do with it.

**Reply by Bibo-Joshi** [reply, marked as answer]
Adding `self` as argument to `wrapper` and passing it along to `func` does the trick for me. Alternatively, you can ofc also just use `*args, **kwargs`

**Reply by klaiderman** [reply]
Did it with `*args, **kwargs`, thanks!

**Comment by klaiderman** [comment]
> Adding `self` as argument to `wrapper` and passing it along to `func` does the trick for me. Alternatively, you can ofc also just use `*args, **kwargs`
